File: src/java/com/portalpostal/dao/handler/LancamentoHandler.java
package com.portalpostal.dao.handler;

import com.portalpostal.model.CentroCusto;
import com.portalpostal.model.Conta;
import com.portalpostal.model.Lancamento;
import com.portalpostal.model.LancamentoProgramado;
import com.portalpostal.model.PlanoConta;
import com.portalpostal.model.dd.TipoLancamento;
import com.portalpostal.model.dd.TipoModeloLancamento;
import com.portalpostal.model.dd.TipoSituacaoLancamento;
import java.sql.ResultSet;
import java.sql.SQLException;
import org.sql2o.ResultSetHandler;

public class LancamentoHandler extends GenericHandler implements ResultSetHandler<Lancamento> {
    
    public LancamentoHandler() {
        super("lancamento");
    }
    
    public LancamentoHandler(String table) {
        super(table);
    }

    public Lancamento handle(ResultSet result) throws SQLException {
        Lancamento lancamento = new Lancamento();
        lancamento.setIdLancamento(getInt(result, "idLancamento"));
        lancamento.setConta(getConta(result));
        lancamento.setPlanoConta(getPlanoConta(result));
        lancamento.setCentroCusto(getCentroCusto(result));
        lancamento.setLancamentoProgramado(getLancamentoProgramado(result));
        lancamento.setTipo(TipoLancamento.values()[getInt(result, "tipo")]);
        lancamento.setFavorecido(getString(result, "favorecido"));
        lancamento.setNumero(getString(result, "numero"));
        lancamento.setNumeroParcela(getInt(result, "numeroParcela"));
        lancamento.setDataCompetencia(getDate(result, "dataCompetencia"));
        lancamento.setDataEmissao(getDate(result, "dataEmissao"));
        lancamento.setDataVencimento(getDate(result, "dataVencimento"));
        lancamento.setDataLancamento(getDate(result, "dataLancamento"));
        lancamento.setDataCompensacao(getDate(result, "dataCompensacao"));
        lancamento.setValor(getDouble(result, "valor"));
        lancamento.setValorDesconto(getDouble(result, "valorDesconto"));
        lancamento.setValorJuros(getDouble(result, "valorJuros"));
        lancamento.setValorMulta(getDouble(result, "valorMulta"));
        lancamento.setSituacao(TipoSituacaoLancamento.values()[getInt(result, "situacao")]);
        lancamento.setModelo(TipoModeloLancamento.values()[getInt(result, "modelo")]);
        lancamento.setNumeroLoteConciliado(getInt(result, "numeroLoteConciliado"));
        lancamento.setAutenticacao(getString(result, "autenticacao"));
        lancamento.setHistorico(getString(result, "historico"));
        lancamento.setObservacao(getString(result, "observacao"));
        lancamento.setUsuario(getString(result, "usuario"));
        lancamento.setAnexos(getAnexos(result));
        return lancamento;
    }
    
    private Conta getConta(ResultSet result) throws SQLException {
        if(!existColumn(result, "conta.idConta")) return null;
        return new ContaHandler().handle(result); 
    }
    
    private PlanoConta getPlanoConta(ResultSet result) throws SQLException {
        if(!existColumn(result, "plano_conta.idPlanoConta")) return null;
        return new PlanoContaHandler().handle(result); 
    }
    
    private CentroCusto getCentroCusto(ResultSet result) throws SQLException {
        if(!existColumn(result, "centro_custo.idCentroCusto")) return null;
        return new CentroCustoHandler().handle(result); 
    }
        
    private LancamentoProgramado getLancamentoProgramado(ResultSet result) throws SQLException {
        if(!existColumn(result, "lancamento_programado.idLancamentoProgramado")) return null;
        if(!existFKValue(result, "lancamento_programado.idLancamentoProgramado")) return null;
        return new LancamentoProgramadoHandler().handle(result); 
    }
        
    private boolean getAnexos(ResultSet result) throws SQLException {
        if(!existColumn(result, "anexos")) return false;
        Long quantidade = getLong(result, "anexos", null);
        if(quantidade > 0) return true;
        return false;
    }
    
}

File: web/app/js/handler/lancamentoHandler.js
'use strict';

app.factory('LancamentoHandler', function(LISTAS) {
            
    var situacoes = LISTAS.situacaoLancamento;
    var modelos = LISTAS.modeloLancamento;
    
    var handleList = function(lancamentoList) { 
        return _.map(lancamentoList, function(lancamento) {
            return handle(lancamento);
        });
    };
    
    var handle = function(lancamento) {  
        var lancamentoHandle = {};
        lancamentoHandle.idLancamento = getId(lancamento);            
        lancamentoHandle.conta = getConta(lancamento);  
        lancamentoHandle.planoConta = getPlanoConta(lancamento);  
        lancamentoHandle.centroCusto = getCentroCusto(lancamento);
        lancamentoHandle.lancamentoProgramado = getLancamentoProgramado(lancamento);
        lancamentoHandle.lancamentoTransferencia = getLancamentoTransferencia(lancamento);
        lancamentoHandle.tipo = getTipo(lancamento); 
        lancamentoHandle.favorecido = getFavorecido(lancamento); 
        lancamentoHandle.numero = getNumero(lancamento); 
        lancamentoHandle.numeroParcela = getNumeroParcela(lancamento); 
        lancamentoHandle.dataCompetencia = getDataCompetencia(lancamento);
        lancamentoHandle.dataEmissao = getDataEmissao(lancamento);
        lancamentoHandle.dataVencimento = getDataVencimento(lancamento);
        lancamentoHandle.dataLancamento = getDataLancamento(lancamento);
        lancamentoHandle.dataCompensacao = getDataCompensacao(lancamento);
        lancamentoHandle.situacao = getSituacao(lancamento);
        lancamentoHandle.modelo = getModelo(lancamento); 
        lancamentoHandle.valor = getValor(lancamento);
        lancamentoHandle.valorDesconto = getValorDesconto(lancamento);
        lancamentoHandle.valorJuros = getValorJuros(lancamento);
        lancamentoHandle.valorMulta = getValorMulta(lancamento);
        lancamentoHandle.numeroLoteConciliado = getNumeroLoteConciliado(lancamento);
        lancamentoHandle.autenticacao = getAutenticacao(lancamento);
        lancamentoHandle.historico = getHistorico(lancamento);
        lancamentoHandle.observacao = getObservacao(lancamento);
        lancamentoHandle.usuario = getUsuario(lancamento);
        lancamentoHandle.anexos = getAnexos(lancamento);
        lancamentoHandle.rateios = null;
        return lancamentoHandle;
    };
    
    var getId = function(lancamento) {
        return lancamento.idLancamento || null;
    };
    
    var getConta = function(lancamento) {
        if(!lancamento.conta) return lancamento.conta;
        return { idConta: lancamento.conta.idConta };
    };
    
    var getPlanoConta = function(lancamento) {
        if(!lancamento.planoConta) return lancamento.planoConta;    
        return { idPlanoConta: lancamento.planoConta.idPlanoConta }; 
    };
    
    var getCentroCusto = function(lancamento) {
        if(!lancamento.centroCusto) return lancamento.centroCusto;    
        return { idCentroCusto: lancamento.centroCusto.idCentroCusto }; 
    };
    
    var getLancamentoProgramado = function(lancamento) {
        if(!lancamento.lancamentoProgramado) return lancamento.lancamentoProgramado;
        return { idLancamentoProgramado: lancamento.lancamentoProgramado.idLancamentoProgramado }; 
    };
    
    var getLancamentoTransferencia = function(lancamento) {
        if(!lancamento.lancamentoTransferencia) return lancamento.lancamentoTransferencia;
        return { idLancamentoTransferencia: lancamento.lancamentoTransferencia.idLancamentoTransferencia }; 
    };
    
    var getTipo = function(lancamento) {
        if(!lancamento.tipo) return lancamento.tipo.id;
        return lancamento.tipo.id; 
    };
    
    var getFavorecido = function(lancamento) {
        return lancamento.favorecido || null;
    };
    
    var getNumero = function(lancamento) {
        return lancamento.numero || null;
    };
    
    var getNumeroParcela = function(lancamento) {
        return lancamento.numeroParcela || null;
    };
    
    var getSituacao = function(lancamento) {
        if(!lancamento.situacao) return lancamento.situacao.id;
        return lancamento.situacao.id; 
    };
    
    var getModelo = function(lancamento) {
        if(!lancamento.modelo) return modelos[0].id;
        return lancamento.modelo.id; 
    };
    
    var getDataCompetencia = function(lancamento) {
        return lancamento.dataCompetencia || moment(); 
    };
    
    var getDataEmissao = function(lancamento) {
        return lancamento.dataEmissao || moment(); 
    };
    
    var getDataVencimento = function(lancamento) {
        return lancamento.dataVencimento || lancamento.dataLancamento || moment(); 
    };
    
    var getDataLancamento = function(lancamento) {
//        return lancamento.dataLancamento || moment(); 
        return lancamento.dataLancamento; 
    };
    
    var getDataCompensacao = function(lancamento) {
        if(lancamento.situacao !== situacoes[2].id) return lancamento.dataCompensacao;
        return lancamento.dataCompensacao = moment();
    };
    
    var getValor = function(lancamento) {
        return lancamento.valor || 0;
    };
    
    var getValorDesconto = function(lancamento) {
        return lancamento.valorDesconto || 0;
    };
    
    var getValorJuros = function(lancamento) {
        return lancamento.valorJuros || 0;
    };
    
    var getValorMulta = function(lancamento) {
        return lancamento.valorMulta || 0;
    };
    
    var getNumeroLoteConciliado = function(lancamento) {
        return lancamento.numeroLoteConciliado || null;
    };
    
    var getAutenticacao = function(lancamento) {
        return lancamento.autenticacao || null;
    };
    
    var getHistorico = function(lancamento) {
        return lancamento.historico || null;
    };
    
    var getObservacao = function(lancamento) {
        return lancamento.observacao || null;
    };
    
    var getUsuario = function(lancamento) {
        return lancamento.usuario || null;
    };
    
    var getAnexos = function(lancamento) {
        return lancamento.anexos || false;
    };
    
    var getDate = function() {
        return moment().format("DD/MM/YYYY")
    }

    return {
        handle: handle,
        handleList: handleList
    }

});